server/unittools.cpp

```diff
@@ -298,12 +298,14 @@ void unit_versus_unit(struct unit *attacker, struct unit *defender,
   player_update_last_war_action(plr1);
   player_update_last_war_action(plr2);
 
-  if (attackpower == 0) {
-    *att_hp = 0;
-  } else if (defensepower == 0) {
-    *def_hp = 0;
-  }
   max_rounds = get_unit_bonus(attacker, EFT_COMBAT_ROUNDS);
+  if (max_rounds <= 0) {
+    if (attackpower == 0 || attack_firepower == 0) {
+      *att_hp = 0;
+    } else if (defensepower == 0 || defense_firepower == 0) {
+      *def_hp = 0;
+    }
+  }
```

Comment:
Is it possible to reach this function if `attack_firepower = defense_firepower = 0` ? I think this would trigger an infinite loop below since `def_hp` and `att_hp` would never be reduced.

Reply:
Surely we'd match the first `if` clause, and set `att_hp = 0`.

Reply:
Oh, right, an infinite loop can only happen if `max_rounds <= 0`.